contracts/token/ERC20/extensions/ERC20uRWA.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import {IERC7943} from "../../../interfaces/IERC7943.sol";
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {ERC165, IERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {ERC20Freezable} from "./ERC20Freezable.sol";
+import {ERC20Restricted} from "./ERC20Restricted.sol";
+
+/**
+ * @dev Extension of {ERC20} according to https://eips.ethereum.org/EIPS/eip-7943[EIP-7943].
+ *
+ * Combines standard ERC-20 functionality with RWA-specific features like user restrictions,
+ * asset freezing, and forced asset transfers.
+ */
+abstract contract ERC20uRWA is ERC20, ERC165, ERC20Freezable, ERC20Restricted, IERC7943 {
+    /// @inheritdoc ERC20Restricted
+    function isUserAllowed(address user) public view virtual override(IERC7943, ERC20Restricted) returns (bool) {
+        return super.isUserAllowed(user);
+    }
+
+    /// @inheritdoc ERC165
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC165, IERC165) returns (bool) {
+        return interfaceId == type(IERC7943).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IERC7943-isTransferAllowed}.
+     *
+     * CAUTION: This function is only meant for external use. Overriding it will not apply the new checks to
+     * the internal {_update} function. Consider overriding {_update} accordingly to keep both functions in sync.
+     */
+    function isTransferAllowed(address from, address to, uint256, uint256 amount) external view virtual returns (bool) {
+        return (amount <= available(from) && isUserAllowed(from) && isUserAllowed(to));
+    }
+
+    /// @inheritdoc IERC7943
+    function getFrozen(address user, uint256) public view virtual returns (uint256 amount) {
+        return frozen(user);
+    }
+
+    /// @inheritdoc IERC7943
+    function setFrozen(address user, uint256, uint256 amount) public virtual {
+        _checkFreezer(user, amount);
+        require(amount <= balanceOf(user), ERC20InsufficientBalance(user, balanceOf(user), amount));
+        _setFrozen(user, amount);
+    }
+
+    /// @inheritdoc IERC7943
+    function forceTransfer(address from, address to, uint256, uint256 amount) public virtual {
+        _checkEnforcer(from, to, amount);
+        require(isUserAllowed(to), ERC7943NotAllowedUser(to));
+
+        // Update frozen balance if needed
+        uint256 currentFrozen = frozen(from);
+        uint256 currentBalance = balanceOf(from);
+        uint256 newAmount = currentBalance - amount;
+        if (currentFrozen > newAmount) {
+            _setFrozen(from, newAmount);
+        }
```

Comment:
Why is this necessary, given we are bypassing the available check?

Reply:
Right. The `if` check is a gas optimization.

1) No if (`setFrozen(Math.min(currentFrozen, newBalance))`): Always executes SSTORE (warm) + emits Frozen event (100 gas + ~1000 gas)
2) With if: Just a JUMPI when no update needed (~10 gas)

Reply:
I'm saying we never really need to call `_setFrozen`